cpp/include/tensorrt_llm/batch_manager/kvCacheManager.h

```diff
@@ -238,6 +270,179 @@ struct KvCacheStats
+    BlockKey mBlockKey;
+    // Flag indicating if block is full
+    bool mIsFull;
+    // Flag indicating if this is root node
+    bool mIsRoot;
+    // Previous node in search structure
+    LookupNodePtr mPrevNode;
+    // Next node(s) in sequence(s)
+    NextNodeMap mNextNodes;
+    // Pointers to blocks holding KV state for this prompt prefix
+    std::unordered_map<SizeType32, BlockPtr> mBlocks;
+};
+
+// Print methods used for debugging.
+std::ostream& operator<<(std::ostream& out, LookupResult const& match);
+std::ostream& operator<<(std::ostream& out, LookupResults const& matches);
+std::ostream& operator<<(std::ostream& out, std::tuple<bool, SizeType32, BlockPtr, LookupNodePtr> const& match);
+std::ostream& operator<<(
+    std::ostream& out, std::vector<std::tuple<bool, SizeType32, BlockPtr, LookupNodePtr>> const& matches);
+std::ostream& operator<<(std::ostream& out,
+    std::unordered_map<SizeType32, std::vector<std::tuple<bool, SizeType32, BlockPtr, LookupNodePtr>>> const& matches);
+
+template <typename T>
+std::string streamPrint(T v)
+{
+    std::stringstream out;
+    out << v;
+    return out.str();
+}
+
+template <typename C, typename... Args>
+std::string streamPrint(C callable, Args... remainingArgs)
+{
+    std::stringstream out;
+    out << callable(out, remainingArgs...);
+    return out.str();
+}
+
+//
+// Basic building block of KV cache.
+//
+// Implements a radix tree that is used to search for blocks containing reusable state for a common prefix.
+// There is only one instance of the search tree, it is owned by BlockManager.
+// Each node in the search tree corresponds to a particular prefix.
+// Each node has a map from window size to KVCacheBlock, if reusable state exists for a given window size,
+// the map will have a pointer to the block containing it.
+// KVCacheBlocks have a pointer back to the node that is referring to them. This makes it easier to remove
+// blocks from the search tree and also allows searches like getPrevBlock().
+//
+// Notes:
+// None of these methods are thread safe. It is up to the caller to ensure thread safety.
+// These methods return values for all window sizes with a single pass through the search tree.
+//
+class KVCachePromptLookup
+{
+public:
+    explicit KVCachePromptLookup(CacheType cacheType, SizeType32 tokensPerBlock);
+
+    //! \brief Return vector of BlockKey for the first inputLength tokens of prompt stored in llmRequest.
+    //! \details If inputLength < 0, effective input length is total number of tokens + inputLength. If you want to skip
+    //! the last token, you can simply do inputLength = -1.
+    [[nodiscard]] std::vector<BlockKey> getBlockKeys(
+        LlmRequest const& llmRequest, SizeType32 inputLength, bool allowPartiallyFilledBlock) const;
+
+    //! \brief Find last valid block for prefix given in blockKey.
+    //! \details Seems like blockKey contains a full prefix, not just key for an individual block. This is very hacky
```

Comment:
Fix comment, multiple `\details`.

I guess it is a design decision upon whether the hash needs to take the full prefix into account since we already operate upon a radix tree. What specific bug do you mean here?